pkg/operator/controller/canary/controller.go

```diff
@@ -306,17 +314,54 @@ func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
 	return nil
 }
 
-func (r *reconciler) setCanaryFailingStatusCondition() error {
+func (r *reconciler) setCanaryFailingStatusCondition(errors []error) error {
+	errorStrings := deduplicateErrorStrings(errors)
+	if len(errorStrings) > canaryFailingNumErrors {
+		errorStrings = errorStrings[len(errorStrings)-canaryFailingNumErrors:]
+	}
 	cond := operatorv1.OperatorCondition{
 		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
 		Status:  operatorv1.ConditionFalse,
 		Reason:  "CanaryChecksRepetitiveFailures",
-		Message: "Canary route checks for the default ingress controller are failing",
+		Message: fmt.Sprintf("Canary route checks for the default ingress controller are failing. Last %d error messages:\n%s", len(errorStrings), strings.Join(errorStrings, "\n")),
 	}
 
 	return r.setCanaryStatusCondition(cond)
 }
 
+// deduplicateErrorStrings takes a slice of errors in chronological order, prunes the list for unique error messages
+// It returns a slice of the error message as strings, with any string seen more than once incuding how many times it
+// was seen
+// The chronological order is preserved, based on the last time each error message was seen
+func deduplicateErrorStrings(errors []error) []string {
+	encountered := map[string]int{}
+	ret := []string{}
+	// Iterate over the list of errors from newest to oldest, keeping track of how many times each error message was
+	// seen. This iteration order makes sure we preserve order based on the newest time each message was seen
+	for i := len(errors) - 1; i >= 0; i-- {
+		if encountered[errors[i].Error()] >= 1 {
+			encountered[errors[i].Error()] += 1
+			continue
+		}
+		// if the error message hasn't been encountered yet, append it to the uniques list. This creates a list in the
+		// reverse order of the initial list, so it will need to be reversed later
+		encountered[errors[i].Error()] = 1
+		ret = append(ret, errors[i].Error())
+	}
+	// Now that all error messages have been tallied, prepend a comment to each message that was seen more than once, so
+	// the reader can tell how often each error has occurred
+	for i := range ret {
+		if encountered[ret[i]] > 1 {
+			ret[i] = fmt.Sprintf("[Seen %d times] %s", encountered[ret[i]], ret[i])
+		}
+	}
+	// Reverse the list so that it can be returned in oldest to newest order
+	for i, j := 0, len(ret)-1; i < j; i, j = i+1, j-1 {
+		ret[i], ret[j] = ret[j], ret[i]
+	}
+	return ret
```

Comment:
We can add to some result slice of the same size as `ret` at the the same time we iterate over `ret`.Something like this should save us 1 run over the slice and the cost of the permutations:
```suggestion
	// Now that all error messages have been tallied, prepend a comment to each message that was seen more than once, so
	// the reader can tell how often each error has occurred
	res, j := make([]string, len(ret)), len(ret)-1
	for i := range ret {
		if encountered[ret[i]] > 1 {
			ret[i] = fmt.Sprintf("[Seen %d times] %s", encountered[ret[i]], ret[i])
		}
		res[j] = ret[i]
		j--
	}
	return res
```